### PR TITLE
#33 efficiency mode

### LINKED ISSUE #33: Implement Efficiency Mode
This is  gonna involve taking the current drive state of the car and translating that to a set of IO behavior to increase the range of the car.

Things to do:
* Potentiometers on the wheel PCB should control the radiator and accumulator fan speeds
* Will need to broadcast a fan speed to shepherd that needs to be interpreted by shepherd
* Paddles change maximum torque (automatically derate the torque calculated based on the broadcasted current limit)
* Implement regen!

### PR DESCRIPTION
**Pull request by Peyton-McKee**
This is all of the elements of efficiency mode other than regen strength which matt says hell do. TBD if done in this ticket or a new one. 

closes #33

### CODE REVIEW

File: MPU/src/gpio.cpp [hunk 1]
```diff
@@ -34,6 +34,23 @@ void GPIO::handlePump()
...
+    }
+
+    if (drive_state == EFFICIENCY) {
+        temp = temp * radiatorFanSpeedPercentage; 
+    }
+    pdu->enableRadiatorFan(temp);
+}
 
-    pdu->enableRadiatorFan(temp > MAX_FANSPEED_TEMP);
+void GPIO::setRadiatorFanPercentage(float speed)
```
Comment (nwdepatie) [inline MPU/src/gpio.cpp:48]: why is this a float ?

File: MPU/src/driverio.cpp [hunk 2]
```diff
@@ -137,4 +159,25 @@ void DriverIO::wheelIO_cb(const CAN_message_t &msg)
...
+    accumulatorFanDial.setDialValue(wheelio.io.pot_l);
+    motorFanDial.setDialValue(wheelio.io.pot_r);
+}
+
+void DriverIO::handleDialState()
+{
+    float motorFanPercentage = ((float)motorFanDial.getDialValue() / 4096.0);
+    if (motorFanPercentage < 0.1) motorFanPercentage = 0.0;
+    else if (motorFanPercentage > 0.9) motorFanPercentage = 1.0;
+    gpio->setRadiatorFanPercentage(motorFanPercentage);
```
Comment (nwdepatie) [inline MPU/src/driverio.cpp:174]: ngl I feel a bit weird calling a GPIO function in here, but I can't think of a better way to do it. GPIO is intended to be some of the "pseudo-controller" classes that handle the logic behind it and intermingling them feels a bit odd, but again I can't really think of a better way to do it lol

File: MPU/include/pedals.h [hunk 3]
```diff
@@ -81,6 +83,20 @@ class Pedals
...
+         * @return int representing the percentage of torque limiting
+        */
+        uint8_t getTorqueLimitPercentage();
+
+        /**
+         * @brief Sets the percentage of torque limiting
+         * 
+         * @param percentage 
+        */
+        void setTorqueLimitPercentage(float percentage);
```
Comment (nwdepatie) [inline MPU/include/pedals.h:99]: echoing again idk why the param is a float, we try and avoid floats at all costs. Can we shift it to be *100 or something to make it not a float? Might be confused on the use too

**Review by nwdepatie — APPROVED** [review]
If its validated in hardware I approve

File: MPU/include/pedals.h (hunk 3 above)
Comment (Peyton-McKee) [reply MPU/include/pedals.h:106]: I initally had it as an int but @mccauleyma said that it doesnt make sense to multiply by 100 just to divide by 100 when we use it

File: MPU/include/gpioHW.h [hunk 4]
```diff
@@ -15,6 +15,8 @@ class PDU
 {
     private:
 
+        uint8_t motor_temp_to_rad_duty[20] = {0, 0, 0, 0, 0, 0, 25, 50, 75, 100, 128, 150, 200, 225, 255, 255, 255, 255, 255, 255};
```
Comment (nwdepatie) [inline MPU/include/gpioHW.h:18]: Is there no better way to do this lol? This feels very odd and unnecessary when we could implement a piecewise curve

File: MPU/include/mpuConfig.h [hunk 5]
```diff
@@ -72,7 +75,14 @@ enum mpu_states
     FAULT,
     MAX_MPU_STATES
 };
-
 extern enum mpu_states mpu_state;
 
+enum Regen_Level_t
+{
+    ZILCH,
```
Comment (nwdepatie) [inline MPU/include/mpuConfig.h:82]: lmao zilch

File: MPU/include/pedals.h [hunk 6]
```diff
@@ -21,9 +21,18 @@
...
 #define BRAKELIGHT_PIN              4
 
+#define MOTOR_RPM_TO_MPH_CONST      0.013048225
+#define PIT_MAX_SPEED               5
+#define ACCUMULATOR_SIZE            5
+
 class Pedals
 {
     private:
+        int REGEN_STRENGTHS[4] = {0, 100, 250, 500}; // N-m to be applied, in N-m * 10
```
Comment (nwdepatie) [inline MPU/include/pedals.h:31]: Why is this a LUT and why are the strengths not linearly going up?

File: MPU/include/pedals.h (hunk 3 above)
Comment (nwdepatie) [reply MPU/include/pedals.h:111]: eh whatever lol

**Review by nwdepatie — APPROVED** [review]
I think that there are cleaner ways to pass messages and also we are using LUT where we could be implementing some curve which feels weird, but I think I'm good with it if it works. We're almost at comp and need stuff merged in

File: MPU/include/gpioHW.h (hunk 4 above)
Comment (Peyton-McKee) [reply MPU/include/gpioHW.h:18]: @mccauleyma

**Pull request merged**